auth/handlers/customers.go

```diff
@@ -0,0 +1,124 @@
+package handlers
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+	"sync"
+	"time"
+)
+
+const customerCacheExpiry = time.Minute * 5
+
+// Customers checks whether users are customers of OpenFaaS Cloud
+type Customers struct {
+	Usernames *map[string]string
+	Sync      *sync.Mutex
+	Expires   time.Time
+}
+
+// NewCustomers creates a Customers instance
+func NewCustomers() *Customers {
+	return &Customers{
+		Sync:    &sync.Mutex{},
+		Expires: time.Now().Add(time.Minute * -1),
+	}
+}
+
+// Get returns whether a customer is found
+func (c *Customers) Get(login string) (bool, error) {
+	found := false
+
+	log.Printf("CUSTOMERS cache expires in: %fs", c.Expires.Sub(time.Now()).Seconds())
+	if c.Expires.Before(time.Now()) {
+		c.Fetch()
+	}
+
+	c.Sync.Lock()
+	lookup := *c.Usernames
+
```

Comment:
If something goes wrong while doing the lookup the sync is never unlocked.

I would do something like:

```go
c.Sync.Lock()
defer c.Sync.Unlock()
lookup := *c.Usernames

if _, ok := lookup[strings.ToLower(login)]; ok {
	return true, nil
}
return false, nil

```

The benefit of this is also that any further change in this code is less likely to brak the unlock logic that is happening anyway.